### PR TITLE
QBFT migration: Create a migrating protocol context

### LINKED ISSUE #3000: Switchable protocol context for consensus migration
Create a protocol context that can switch and delegate to the appropriate protocol context object needed for a consensus mechanism at a particular height. This is needed so that the context object is the correct one for the current block chain height when different consensus mechanisms can be used at different heights.

**Acceptance Criteria**
* Protocol context returns the context object for the consensus mechanism at the current height.

### PR DESCRIPTION
**Pull request by jframe**
## PR description
For the QBFT migration we need to be able to return the appropriate consensus context for the current consensus mechanism.

This introduces a schedulable protocol context and scheduable context that are used together to return the consensus context for the current chain height.

Fixes #3000 
## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: consensus/common/src/main/java/org/hyperledger/besu/consensus/common/ScheduableProtocolContext.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+  public static ProtocolContext init(
+      final MutableBlockchain blockchain,
+      final WorldStateArchive worldStateArchive,
+      final GenesisState genesisState,
+      final ProtocolSchedule protocolSchedule,
+      final ConsensusContextFactory consensusContextFactory) {
+    if (blockchain.getChainHeadBlockNumber() < 1) {
+      genesisState.writeStateTo(worldStateArchive.getMutable());
+    }
```
Comment (siladu) [inline consensus/common/src/main/java/org/hyperledger/besu/consensus/common/ScheduableProtocolContext.java:46]: Is this actually related to the ProtocolContext? I'm wondering why this is here and not in BesuController?

I see it's come from ProtocolContext.init and so my query is not really relevant to this PR, although arguably this is a good moment to resolve it if we're about to duplicate it.

File: consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/BftForksSchedule.java [hunk 2]
```diff
@@ -31,7 +31,7 @@
 
 import com.google.common.annotations.VisibleForTesting;
 
-public class BftForksSchedule<C extends BftConfigOptions> {
+public class BftForksSchedule<C> {
```
Comment (siladu) [inline consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/BftForksSchedule.java:34]: Can merge the ForksSchedule refactored code from main now, as I'm sure you know :)

File: besu/src/main/java/org/hyperledger/besu/controller/ConsensusScheduleBesuControllerBuilder.java [hunk 3]
```diff
@@ -135,14 +141,38 @@ protected ProtocolSchedule createProtocolSchedule() {
...
+    final List<ForkSpec<ConsensusContext>> consensusContextSpecs =
+        besuControllerBuilderSchedule.entrySet().stream()
+            .map(
+                e ->
+                    new ForkSpec<>(
+                        e.getKey(),
+                        e.getValue()
+                            .createConsensusContext(
+                                blockchain, worldStateArchive, protocolSchedule)))
+            .sorted(ForkSpec.COMPARATOR)
```
Comment (siladu) [inline besu/src/main/java/org/hyperledger/besu/controller/ConsensusScheduleBesuControllerBuilder.java:169]: ah nice, this will be useful for the mining coordinator too :)

File: besu/src/test/java/org/hyperledger/besu/controller/ConsensusScheduleBesuControllerBuilderTest.java [hunk 4]
```diff
@@ -98,4 +105,43 @@ public void mustCreateCombinedProtocolScheduleUsingProtocolSchedulesOrderedByBlo
...
   }
+
+  @Test
+  public void mustCreateSchedulableContext() {
+    final ConsensusContext context1 = Mockito.mock(ConsensusContext.class);
+    final ConsensusContext context2 = Mockito.mock(ConsensusContext.class);
+
+    final Map<Long, BesuControllerBuilder> besuControllerBuilderSchedule = new TreeMap<>();
+    besuControllerBuilderSchedule.put(0L, besuControllerBuilder1);
+    besuControllerBuilderSchedule.put(10L, besuControllerBuilder2);
```
Comment (siladu) [inline besu/src/test/java/org/hyperledger/besu/controller/ConsensusScheduleBesuControllerBuilderTest.java:116]: Is it worth using a HashMap or something and making these out of order to test the `.sorted(ForkSpec.COMPATOR)`?

File: consensus/common/src/main/java/org/hyperledger/besu/consensus/common/ScheduableProtocolContext.java (hunk 1 above)
Comment (siladu) [reply consensus/common/src/main/java/org/hyperledger/besu/consensus/common/ScheduableProtocolContext.java:46]: This also wants coverage according to Sonarcloud...not sure if we've agreed to adhere to a sensible coverage limit yet though

**Review by siladu — COMMENTED** [review]
Couple of comments and questions, no biggies.

What are your thoughts on keeping Sonarcloud happy w.r.t unit test coverage?

File: besu/src/test/java/org/hyperledger/besu/controller/ConsensusScheduleBesuControllerBuilderTest.java (hunk 4 above)
Comment (siladu) [reply besu/src/test/java/org/hyperledger/besu/controller/ConsensusScheduleBesuControllerBuilderTest.java:116]: Having investigated doing this for https://github.com/hyperledger/besu/pull/3097#discussion_r762392580
I don't think it's possible due to the way ForksSchedule is implemented

File: consensus/common/src/main/java/org/hyperledger/besu/consensus/common/ScheduableProtocolContext.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import org.hyperledger.besu.consensus.common.bft.BftForksSchedule;
+import org.hyperledger.besu.ethereum.ConsensusContext;
+import org.hyperledger.besu.ethereum.ConsensusContextFactory;
+import org.hyperledger.besu.ethereum.ProtocolContext;
+import org.hyperledger.besu.ethereum.chain.GenesisState;
+import org.hyperledger.besu.ethereum.chain.MutableBlockchain;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSchedule;
+import org.hyperledger.besu.ethereum.worldstate.WorldStateArchive;
+
+public class ScheduableProtocolContext extends ProtocolContext {
```
[the remark this answers is not in the document]
Comment (jframe) [reply consensus/common/src/main/java/org/hyperledger/besu/consensus/common/ScheduableProtocolContext.java:25]: Still not convinced on it either. Thinking maybe MigratingMiningCoordinator but also want them to align.

File: consensus/common/src/main/java/org/hyperledger/besu/consensus/common/ScheduableProtocolContext.java (hunk 1 above)
Comment (jframe) [reply consensus/common/src/main/java/org/hyperledger/besu/consensus/common/ScheduableProtocolContext.java:45]: Don't know why it's there don't seem a bit weird now I think about it. But yes copied it from the ProtocolContext.init method.

File: besu/src/main/java/org/hyperledger/besu/controller/ConsensusScheduleBesuControllerBuilder.java (hunk 3 above)
Comment (jframe) [reply besu/src/main/java/org/hyperledger/besu/controller/ConsensusScheduleBesuControllerBuilder.java:169]: Think I can ditch the sorted now based on #3097

**Review by usmansaleem — APPROVED** [review]
LGTM

File: consensus/common/src/main/java/org/hyperledger/besu/consensus/common/ScheduableProtocolContext.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jframe) [reply consensus/common/src/main/java/org/hyperledger/besu/consensus/common/ScheduableProtocolContext.java]: Renamed to MigratingProtocolContext

File: besu/src/test/java/org/hyperledger/besu/controller/ConsensusScheduleBesuControllerBuilderTest.java [hunk 6]
```diff
@@ -162,4 +167,43 @@ public void createsScheduableMiningCoordinator() {
               .isSameAs(miningCoordinator2);
         });
   }
+
+  @Test
+  public void mustCreateSchedulableContext() {
```
Comment (siladu) [inline besu/src/test/java/org/hyperledger/besu/controller/ConsensusScheduleBesuControllerBuilderTest.java:172]: rename to `mustCreateMigratingContext`?

**Review by siladu — APPROVED** [review]
